src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1259,52 +1284,107 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
-                // Win32FileStream's ReadAsync blocks waiting for data to fill it's 
-                // internal buffer, you will be blocked.  In a case where a child
-                // process writes to stdout & stderr while a parent process tries
-                // reading from both, you can easily get into a deadlock here.
-                // To avoid this deadlock, don't buffer when doing async IO on
-                // pipes.   
-                Debug.Assert(_readPos == 0 && _readLen == 0, "Win32FileStream must not have buffered data here!  Pipes should be unidirectional.");
+                // Ensure the buffer is clear for writing
+                if (_writePos == 0)
+                {
+                    if (_readPos < _readLen)
+                    {
+                        FlushRead();
+                    }
+                    _readPos = 0;
+                    _readLen = 0;
+                }
 
-                if (_writePos > 0)
-                    FlushWrite(false);
+                // Determine how much space remains in the buffer
+                int remainingBuffer = _bufferSize - _writePos;
+                Debug.Assert(remainingBuffer >= 0);
 
-                return WriteInternalCoreAsync(array, offset, numBytes, cancellationToken);
-            }
+                // Simple/common case:
+                // - The write is smaller than our buffer, such that it's worth considering buffering it.t
+                // - There's no active flush operation, such that we don't have to worry about the existing buffer being in use.
+                // - And the data we're trying to write fits in the buffer, meaning it wasn't already filled by previous writes.
+                // In that case, just store it in the buffer.
+                if (numBytes < _bufferSize && !HasActiveBufferOperation && numBytes <= remainingBuffer)
+                {
+                    if (_buffer == null)
+                        _buffer = new byte[_bufferSize];
 
-            // Handle buffering.
-            if (_writePos == 0)
-            {
-                if (_readPos < _readLen) FlushRead();
-                _readPos = 0;
-                _readLen = 0;
+                    Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
+                    _writePos += numBytes;
+                    writeDataStoredInBuffer = true;
+
+                    // There is one special-but-common case (common because devs often use
+                    // byte[] sizes that are powers of 2 and thus fit nicely into our buffer, which is
+                    // also a power of 2), where we fill the buffer to the brim. In that case, we want 
+                    // to issue the asynchronous flush/write now rather than waiting until the next write.
+                    // So we only return a completed task here rather than continuing on to do the write
+                    // if the buffer isn't yet full.
```

Comment:
I'm not certain that I understand the comment here. At first it says that there is a special-but-common case where we fill the buffer to the brim and want to issue a flush/write now. but then it says we return a completed task rather than writing if the buffer isn't full.

It seems like the code is actually handling the <i>opposite</i> case from the comment i.e. the buffer is <i>not</i> filled to the brim. I think I understand now what the comment is saying, but a reword would certainly be worthwhile :)

Something like:

```
// There is one special-but-common case (common because devs often use
// byte[] sizes that are powers of 2 and thus fit nicely into our buffer, which is
// also a power of 2), where we fill the buffer to the brim. In that case, we want 
// to issue the asynchronous flush/write now rather than waiting until the next write.

// However, if the buffer still has remaining space for future writes we instead return a
// Task.CompletedTask and defer the write to a later operation.